api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,254 @@
+  google.protobuf.Struct dynamic_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // Information about the client's connection -- downstream if this is
+  // a request and upstream if it is a response.
+  auth.v3.AttributeContext.Peer client_info = 2;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 3;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 4;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a request_headers or response_headers
+// message.
+message HeadersResponse {
+  // Instructions on how to manipulate the headers
+  HeaderMutation header_mutation = 1;
+
+  // If specified, add a body to the message if none was present
+  // before.
+  BodyMutation body_mutation = 2;
+}
+
+// This message must be sent in response to a request_trailers
+// or response_trailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the trailers
+  HeaderMutation header_mutation = 1;
+}
+
+message BodyResponse {
+  // If set, manipulate HTTP headers along with the body. Note that
+  // this is only possible if the headers have not been sent down
+  // the filter chain yet. Practically, then, this field will be ignored
+  // unless the request or response BodySendMode is BUFFERED or
+  // BUFFERED_PARTIAL, which means that Envoy is buffering the body
+  // in memory while waiting for this response.
+  HeaderMutation header_mutation = 1;
+
+  // Request that the last body chunk delivered be modified.
+  BodyMutation body_mutation = 2;
+}
+
+// This message completely replaces the response.
+// It will immediately return the data specified here to the downstream
+// system and stop further filter processing.
```

Comment:
I would clarify what happens if the response has already been started when this is received.